sql/wsrep_server_service.cc

```diff
@@ -302,3 +302,78 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
 void Wsrep_server_service::debug_sync(const char*)
 {
 }
+
+// IV vector is the size of a block, which for AES is 16
+const unsigned int encryption_iv_size = MY_AES_BLOCK_SIZE;
+
+int Wsrep_server_service::do_crypt(void**                ctx,
+                                   wsrep::const_buffer&  key,
+                                   unsigned int          key_version,
+                                   const char            (*iv)[32],
+                                   wsrep::const_buffer&  input,
+                                   void*                 output,
+                                   bool                  encrypt,
+                                   bool                  last)
+{
+  if (*ctx == NULL)
+  {
+    int mode= encrypt ? ENCRYPTION_FLAG_ENCRYPT : ENCRYPTION_FLAG_DECRYPT;
+    *ctx = ::malloc(encryption_ctx_size(ENCRYPTION_KEY_SYSTEM_DATA,
+                                        key_version));
+    if (*ctx == NULL)
+    {
+      throw wsrep::runtime_error("Memory not allocated in do_crypt()");
+    }
+    if(encryption_ctx_init(*ctx, (const unsigned char*)key.data(), key.size(),
+                           (unsigned char*)iv, encryption_iv_size,
+                           mode | ENCRYPTION_FLAG_NOPAD,
+                           ENCRYPTION_KEY_SYSTEM_DATA,
+                           key_version))
+    {
+      throw wsrep::runtime_error("Failed encryption_ctx_init()");
+    }
+  }
+  unsigned int ctx_update_size= 0;
+  if(encryption_ctx_update(*ctx, (const unsigned char*)input.data(),
+                           input.size(),
+                           (unsigned char *)output, &ctx_update_size))
+  {
+    throw wsrep::runtime_error("Failed encryption_ctx_update()");
+  }
+  unsigned int ctx_finish_size= 0;
+  if (last)
+  {
+    if(encryption_ctx_finish(*ctx, (unsigned char *)output + ctx_update_size,
+                              &ctx_finish_size))
+    {
+      throw wsrep::runtime_error("Failed encryption_ctx_finish()");
+    }
+    assert(ctx_update_size + ctx_finish_size == input.size());
+    free(*ctx);
+  }
+
+  return ctx_update_size + ctx_finish_size;
+}
+
+const unsigned int key_version_preamble_size = 8;
+
+void Wsrep_server_service::key_serialize(std::vector<unsigned char>& output,
+                                         const void* key,
+                                         size_t size,
+                                         unsigned int version)
+{
+  output.resize(key_version_preamble_size + size);
+  memcpy(output.data(), (void *)&version, sizeof(version));
+  memcpy(output.data() + key_version_preamble_size, key, size);
+}
+
+void Wsrep_server_service::key_deserialize(void const* input,
+                                           const size_t& input_size,
```

Comment:
And passing size_t by reference here is perfectly fine, but a bit misleading. In total the signature of this method implies that both contents of `input` buffer and its length can be changed